openssl/src/ssl/mod.rs

```diff
@@ -886,6 +976,29 @@ impl Ssl {
             SslMethod::from_raw(method)
         }
     }
+
+    /// Returns the server's name for the current connection
+    pub fn get_servername(&self) -> Option<String> {
+        let name = unsafe { ffi::SSL_get_servername(self.ssl, ffi::TLSEXT_NAMETYPE_host_name) };
+        if name == ptr::null() {
+            return None;
+        }
+
+        unsafe {
+            String::from_utf8(CStr::from_ptr(name).to_bytes().to_vec()).ok()
+        }
+    }
+
+    /// change the context corresponding to the current connection
+    pub fn set_ssl_context(&self, ctx: &SslContext) -> SslContext {
+        SslContext { ctx: unsafe { ffi::SSL_set_SSL_CTX(self.ssl, ctx.ctx) } }
+    }
+
+    /// obtain the context corresponding to the current connection
+    pub fn get_ssl_context(&self) -> SslContext {
+        let ssl_ctx = unsafe { ffi::SSL_get_SSL_CTX(self.ssl) };
+        SslContext { ctx: ssl_ctx }
```

Comment:
Unfortunately, same issue as when `Ssl` is dropped. The clean solution would be to increase the refcount of SSL_CTX, but since it is an opaque structure, it is not possible from Rust. Maybe in a shim?